datascience_eda/datascience_eda.py

```diff
@@ -74,31 +168,293 @@ def explore_clustering(
 
     Examples
     -------
-    >>> explore_clustering(X)
+    >>> original_df = pd.read_csv(currentdir + "/data/menu.csv")
+    >>> numeric_features = eda.get_numeric_columns(original_df)
+    >>> drop_features = []
+    >>> numeric_transformer = make_pipeline(SimpleImputer(), StandardScaler())
+    >>> preprocessor = make_column_transformer(
+    >>>     (numeric_transformer, numeric_features), ("drop", drop_features)
+    >>> )
+    >>> df = pd.DataFrame(
+    >>>     data=preprocessor.fit_transform(original_df), columns=numeric_features
+    >>> )
+    >>> explore_clustering(df)
     """
+
+    # region validate parameters, throw exception upon invalid ones
+
+    if not (type(df) == pd.DataFrame):
+        raise TypeError("df must be a DataFrame.")
+
+    all_num_cols = get_numeric_columns(df)
+    if numeric_cols is None:
+        numeric_cols = all_num_cols
+    else:
+        # check if the column names are valid
+        _verify_numeric_cols(df, numeric_cols)
+
+    if not (type(hyperparameter_dict) == dict):
+        raise TypeError("hyperparameter_dict must be a dict.")
+
+    if not ("KMeans" in hyperparameter_dict):
+        raise Exception("Expecting Kmeans hyperparams.")
+
+    if not ("DBSCAN" in hyperparameter_dict):
+        raise Exception("Expecting DBSCAN hyperparams.")
+
+    kmeans_params = hyperparameter_dict["KMeans"]
+    if not ("n_clusters" in kmeans_params):
+        raise Exception("Expecting n_clusters in KMeans' hyperparams.")
+
+    dbscan_params = hyperparameter_dict["DBSCAN"]
+    if not ("eps" in dbscan_params):
+        raise Exception("Excepting eps in DBSCAN's hyperparams.")
+
+    if not ("min_samples" in dbscan_params):
+        raise Exception("Expecting min_samples in DBSCAN's hyperparams.")
+
+    if not ("distance_metric" in dbscan_params):
+        raise Exception("Expecting distance_metric as a hyperparameter")
+
+    metric = dbscan_params["distance_metric"]
+    _verify_distance_metric(metric)
+
+    # endregion
+
+    print("***********************")
+    print("EXPLORE CLUSTERING")
+    print("***********************")
+
+    kmeans_plots = explore_KMeans_clustering(
+        df,
+        num_cols=numeric_cols,
+        n_clusters=kmeans_params["n_clusters"],
+    )
+
+    dbscan_plots = explore_DBSCAN_clustering(
```

Comment:
explore_DBSCAN_clustering is defined after explore_clustering. How would it be if explore_DBSCAN_clustering is defined before explore_clustering so that there would be a flow while referring to the code?